packages/swingset-liveslots/src/virtualObjectManager.js

```diff
@@ -586,273 +661,282 @@ export function makeVirtualObjectManager(
+
+    const contextCache = makeContextCache(makeState, makeContext);
+    allCaches.push(contextCache);
+
+    // defendPrototype/defendPrototypeKit accept a contextProvider function,
+    // or a contextProviderKit record which maps facet name strings to
+    // provider functions. It calls the function during invocation of each
+    // method, and expects to get back the "context" record, either { state,
+    // self } for single-facet VOs, or { state, facets } for multi-facet
+    // ones. The provider we use fetches the state data (if not already in the
+    // cache) at the last minute. This moves any syscalls needed by
+    // stateCache.get() out of deserialization time (which is sensitive to GC)
+    // and into method-invocation time (which is not).
+
+    let proto;
+    if (multifaceted) {
+      proto = defendPrototypeKit(
+        tag,
+        makeContextProviderKit(contextCache, getSlotForVal, facetNames),
+        behavior,
+        thisfulMethods,
+        interfaceGuard,
+      );
+    } else {
+      proto = defendPrototype(
+        tag,
+        makeContextProvider(contextCache, getSlotForVal),
+        behavior,
+        thisfulMethods,
+        interfaceGuard,
+      );
+    }
+    harden(proto);
+
+    // this builds new Representatives, both when creating a new instance and
+    // for reanimating an existing one when the old rep gets GCed
+
+    function reanimateVO(_baseRef) {
+      if (multifaceted) {
+        return makeFacets(facetNames, proto, linkToCohort, unweakable);
+      } else {
+        return makeRepresentative(proto);
+      }
     }
 
     function deleteStoredVO(baseRef) {
       let doMoreGC = false;
-      const rawState = fetch(baseRef);
-      if (rawState) {
-        for (const propValue of Object.values(rawState)) {
-          propValue.slots.forEach(vref => {
-            doMoreGC = vrm.removeReachableVref(vref) || doMoreGC;
-          });
-        }
+      const record = dataCache.get(baseRef);
+      for (const valueCD of Object.values(record.capdatas)) {
+        valueCD.slots.forEach(vref => {
+          doMoreGC = vrm.removeReachableVref(vref) || doMoreGC;
+        });
       }
-      syscall.vatstoreDelete(`vom.${baseRef}`);
+      dataCache.delete(baseRef);
       return doMoreGC;
     }
 
+    // Tell the VRM about this Kind.
+    vrm.registerKind(kindID, reanimateVO, deleteStoredVO, isDurable);
+    vrm.rememberFacetNames(kindID, facetNames);
+
     function makeNewInstance(...args) {
       const id = getNextInstanceID(kindID, isDurable);
       const baseRef = makeBaseRef(kindID, id, isDurable);
       // kdebug(`vo make ${baseRef}`);
 
       const initialData = init ? init(...args) : {};
-      const rawState = {};
+      if (typeof initialData !== 'object') {
+        Fail`initial data must be object, not ${initialData}`;
+        // a common mistake is to use `() => {foo:1}`, not `() => ({foo:1})`
+      }
```

Comment:
```suggestion
      // catch mistaken use of `() => { foo: 1 }` rather than `() => ({ foo: 1 })`
      // (the former being a function with a body having a no-op statement labeled
      // "foo" and returning undefined, the latter being a function with a concise
      // body that returns an object having a property named "foo").
      typeof initialData === 'object' ||
        Fail`initial data must be object, not ${initialData}`;
```